### PR TITLE
enabler - add RW lock for index spec - MOD-4567

### PR DESCRIPTION
**Pull request by DvirDukhan**
This PR is an enabler for locking index spec to read/write
It introduced the capability to lock the spec via the RedisSearchCtx object. 
It should be followed by PRs that validate the usage of this capability in index read/write and GC

### CODE REVIEW

File: src/inverted_index.c [hunk 1]
```diff
@@ -104,10 +104,10 @@ static void IR_SetAtEnd(IndexReader *r, int value) {
 void IndexReader_OnReopen(void *privdata) {
 
   IndexReader *ir = privdata;
+  RedisSearchCtx sctx = SEARCH_CTX_STATIC(RSDummyContext, (IndexSpec *)ir->sp);
   if (ir->record->type == RSResultType_Term) {
     // we need to reopen the inverted index to make sure its stil valid.
     // the GC might have deleted it by now.
-    RedisSearchCtx sctx = (RedisSearchCtx)SEARCH_CTX_STATIC(RSDummyContext, (IndexSpec *)ir->sp);
     InvertedIndex *idx = Redis_OpenInvertedIndexEx(&sctx, ir->record->term.term->str,
                                                    ir->record->term.term->len, 0, NULL, NULL);
```
Comment (GuyAv46) [inline src/inverted_index.c:112]: Why was this line moved outside the condition check? Also, should we pass the `IndexSpec` to the function instead of a search context?

File: src/spec.c [hunk 2]
```diff
@@ -1501,6 +1503,19 @@ IndexSpec *NewIndexSpec(const char *name) {
...
+	res = pthread_rwlockattr_init(&attr) ;
+  RedisModule_Assert(res == 0);
+#if !defined(__APPLE__) && !defined(__FreeBSD__)
+	int pref = PTHREAD_RWLOCK_PREFER_WRITER_NONRECURSIVE_NP ;
+	res = pthread_rwlockattr_setkind_np(&attr, pref) ;
+	RedisModule_Assert(res == 0);
+#endif
+
+	RedisModule_Assert(res == 0);
+  pthread_rwlock_init(&sp->rwlock, &attr);
```
Comment (GuyAv46) [inline src/spec.c:1518]:
```suggestion

  int res = 0;
  pthread_rwlockattr_t attr;
  res = pthread_rwlockattr_init(&attr);
  RedisModule_Assert(res == 0);
#if !defined(__APPLE__) && !defined(__FreeBSD__)
  int pref = PTHREAD_RWLOCK_PREFER_WRITER_NONRECURSIVE_NP;
  res = pthread_rwlockattr_setkind_np(&attr, pref);
  RedisModule_Assert(res == 0);
#endif

  RedisModule_Assert(res == 0);
  pthread_rwlock_init(&sp->rwlock, &attr);
```

File: src/spec.c [hunk 3]
```diff
@@ -1275,6 +1275,8 @@ void IndexSpec_FreeInternals(IndexSpec *spec) {
   } else {
     thpool_add_work(cleanPool, (thpool_proc)IndexSpec_FreeUnlinkedData, spec);
   }
+
+  pthread_rwlock_destroy(&spec->rwlock);
```
Comment (GuyAv46) [inline src/spec.c:1279]: How or where do we validate that the lock is not locked?

File: tests/cpptests/test_cpp_forkgc.cpp [hunk 4]
```diff
@@ -142,7 +141,9 @@ static std::string numToDocid(unsigned id) {
 TEST_F(FGCTest, testRepairLastBlockWhileRemovingMiddle) {
   // Delete the first block:
   unsigned curId = 0;
-  auto iv = getTagInvidx(ctx, sp, "f1", "hello");
+  RedisSearchCtx sctx = SEARCH_CTX_STATIC(ctx, sp);
+  RedisSearchCtx_UnlockSpec(&sctx);
```
Comment (GuyAv46) [inline tests/cpptests/test_cpp_forkgc.cpp:145]: why?

File: tests/cpptests/test_cpp_forkgc.cpp [hunk 5]
```diff
@@ -185,7 +186,9 @@ TEST_F(FGCTest, testRepairLastBlockWhileRemovingMiddle) {
 TEST_F(FGCTest, testRepairLastBlock) {
   // Delete the first block:
   unsigned curId = 0;
-  auto iv = getTagInvidx(ctx, sp, "f1", "hello");
+  RedisSearchCtx sctx = SEARCH_CTX_STATIC(ctx, sp);
+  RedisSearchCtx_UnlockSpec(&sctx);
```
Comment (GuyAv46) [inline tests/cpptests/test_cpp_forkgc.cpp:190]: why?

File: tests/cpptests/test_cpp_forkgc.cpp [hunk 6]
```diff
@@ -222,7 +225,9 @@ TEST_F(FGCTest, testRepairLastBlock) {
 TEST_F(FGCTest, testRepairMiddleRemoveLast) {
   // Delete the first block:
   unsigned curId = 0;
-  auto iv = getTagInvidx(ctx, sp, "f1", "hello");
+  RedisSearchCtx sctx = SEARCH_CTX_STATIC(ctx, sp);
+  RedisSearchCtx_UnlockSpec(&sctx);
```
Comment (GuyAv46) [inline tests/cpptests/test_cpp_forkgc.cpp:229]: why?

File: tests/cpptests/test_cpp_forkgc.cpp [hunk 7]
```diff
@@ -261,7 +266,9 @@ TEST_F(FGCTest, testRepairMiddleRemoveLast) {
 TEST_F(FGCTest, testRemoveMiddleBlock) {
   // Delete the first block:
   unsigned curId = 0;
-  InvertedIndex *iv = getTagInvidx(ctx, sp, "f1", "hello");
+  RedisSearchCtx sctx = SEARCH_CTX_STATIC(ctx, sp);
+  RedisSearchCtx_UnlockSpec(&sctx);
```
Comment (GuyAv46) [inline tests/cpptests/test_cpp_forkgc.cpp:270]: why?

File: src/redis_index.c [hunk 8]
```diff
@@ -163,6 +163,30 @@ RedisModuleString *fmtRedisScoreIndexKey(RedisSearchCtx *ctx, const char *term,
                                         (int)len, term);
 }
 
+void RedisSearchCtx_LockSpecRead(RedisSearchCtx *ctx) {
+  RedisModule_Assert(ctx->flags == RS_CTX_UNSET && ctx->isLocked == false);
```
Comment (oshadmi) [inline src/redis_index.c:167]: Failing an assert if `ctx` is already locked?

File: src/search_ctx.h [hunk 9]
```diff
@@ -23,43 +23,44 @@ extern "C" {
...
   RedisModuleCtx *redisCtx;
   RedisModuleKey *key_;
   IndexSpec *spec;
-  uint32_t refcount;
-  int isStatic;
   uint64_t specId;  // Unique id of the spec; used when refreshing
   struct timespec timeout;
   unsigned int apiVersion; // API Version to allow for backward compatibility / alternative functionality
+  RSContextFlags flags;
+  bool isLocked;
```
Comment (oshadmi) [inline src/search_ctx.h:42]: Is `isLocked` redundant?
(`flags` == `RS_CTX_UNSET` actually means also `isLocked` == 0 ?)

File: src/spec.c [hunk 10]
```diff
@@ -1501,6 +1503,19 @@ IndexSpec *NewIndexSpec(const char *name) {
...
   memset(&sp->stats, 0, sizeof(sp->stats));
+  
+  int res = 0;
+  pthread_rwlockattr_t attr;
+	res = pthread_rwlockattr_init(&attr) ;
+  RedisModule_Assert(res == 0);
+#if !defined(__APPLE__) && !defined(__FreeBSD__)
+	int pref = PTHREAD_RWLOCK_PREFER_WRITER_NONRECURSIVE_NP ;
+	res = pthread_rwlockattr_setkind_np(&attr, pref) ;
+	RedisModule_Assert(res == 0);
```
Comment (oshadmi) [inline src/spec.c:1514]: Fix indentation

File: src/spec.c [hunk 11]
```diff
@@ -1501,6 +1503,19 @@ IndexSpec *NewIndexSpec(const char *name) {
...
+#endif
+
+	RedisModule_Assert(res == 0);
```
Comment (oshadmi) [inline src/spec.c:1517]: Redundant line?

File: src/spec.c [hunk 12]
```diff
@@ -1501,6 +1503,19 @@ IndexSpec *NewIndexSpec(const char *name) {
   sp->used_dialects = 0;
 
   memset(&sp->stats, 0, sizeof(sp->stats));
+  
+  int res = 0;
+  pthread_rwlockattr_t attr;
+	res = pthread_rwlockattr_init(&attr) ;
+  RedisModule_Assert(res == 0);
+#if !defined(__APPLE__) && !defined(__FreeBSD__)
+	int pref = PTHREAD_RWLOCK_PREFER_WRITER_NONRECURSIVE_NP ;
```
Comment (oshadmi) [inline src/spec.c:1512]: What happens on APPLE or FreeBSD - can Writers be starved?

File: src/inverted_index.c (hunk 1 above)
Comment (DvirDukhan) [reply src/inverted_index.c:112]: we need the `RedisSearchCtx` in `Redis_OpenInvertedIndexEx`

File: src/spec.c (hunk 3 above)
Comment (DvirDukhan) [reply src/spec.c:1279]: see https://stackoverflow.com/questions/55614415/pthread-rwlock-destroy-of-a-locked-mutex
the plan is to have the spec ref counted. will explain in person

File: tests/cpptests/test_cpp_forkgc.cpp (hunk 5 above)
Comment (DvirDukhan) [reply tests/cpptests/test_cpp_forkgc.cpp:190]: leftovers

File: src/redis_index.c (hunk 8 above)
Comment (DvirDukhan) [reply src/redis_index.c:167]: No re-entrence is allowed

File: src/search_ctx.h (hunk 9 above)
Comment (DvirDukhan) [reply src/search_ctx.h:42]: yes

File: src/spec.c (hunk 12 above)
Comment (DvirDukhan) [reply src/spec.c]: I took this one from RedisGraph. We need to double-check this, but the answer is yes

**Review by oshadmi — APPROVED** [review]
LGTM 👍🏼
